server/src/main/java/io/crate/replication/logical/ShardReplicationService.java

```diff
@@ -0,0 +1,135 @@
+                                   LogicalReplicationService logicalReplicationService,
+                                   LogicalReplicationSettings replicationSettings,
+                                   Client client,
+                                   ThreadPool threadPool) {
+        this.settings = settings;
+        this.logicalReplicationService = logicalReplicationService;
+        this.replicationSettings = replicationSettings;
+        this.client = client;
+        this.threadPool = threadPool;
+        indexEventListenerProxy.addLast(new LifecycleListener());
+    }
+
+    @Override
+    public void close() throws IOException {
+        for (var tracker : shards.values()) {
+            tracker.close();
+        }
+    }
+
+    @Nullable
+    private Client getRemoteClusterClient(String indexName) {
+        var subscriptionName = logicalReplicationService.subscriptionName(indexName);
+        if (subscriptionName != null) {
+            return logicalReplicationService.getRemoteClusterClient(threadPool, subscriptionName);
+        }
+        return null;
+    }
+
+    private class LifecycleListener implements IndexEventListener {
+
+        @Override
+        public void afterIndexShardStarted(IndexShard indexShard) {
+            if (indexShard.routingEntry().primary()) {
+                var subscription = logicalReplicationService.subscription(indexShard.shardId().getIndexName());
+                if (subscription != null) {
+                    var tracker = new ShardReplicationChangesTracker(
+                        indexShard,
+                        replicationSettings,
+                        threadPool,
+                        ShardReplicationService.this::getRemoteClusterClient,
+                        client
+                    );
+                    shards.put(indexShard.shardId(), tracker);
+                    tracker.start();
+                }
+            }
+        }
+
+        @Override
+        public void beforeIndexShardClosed(ShardId shardId,
+                                           @Nullable IndexShard indexShard,
+                                           Settings indexSettings) {
+            if (logicalReplicationService.isSubscribedIndex(shardId.getIndexName())) {
+                var tracker = shards.remove(shardId);
+                if (tracker != null) {
+                    LOGGER.debug("[{}] Closing shard changes tracker on shard close", shardId);
+                    try {
+                        tracker.close();
+                    } catch (IOException e) {
+                        LOGGER.debug("Error while closing shard changes tracker of shardId=" + shardId, e);
+                    }
+                }
+            }
+        }
+
+        @Override
+        public void afterIndexShardDeleted(ShardId shardId, Settings indexSettings) {
+            if (logicalReplicationService.isSubscribedIndex(shardId.getIndexName())) {
+                var tracker = shards.remove(shardId);
```

Comment:
Couldn't it be the case that the subscription is removed after `afterIndexShardStarted` - and then the shard is closed or deleted?

Looks like we'd never remove the `shards` entry?